### PR TITLE
[AMDGPU] Vectorize i8 Shuffles

### PR DESCRIPTION
**Pull request by jrbyrnes**
i8 shuffle vectors have standard isel dags and will be lowered into v_perms with CalculateByteProvider. 

Moreever, if we are unconvinced by https://github.com/llvm/llvm-project/pull/95328 , enabling i8 shuffle vectorization should indirectly capture many of the needed vectorization cases for https://github.com/llvm/llvm-project/pull/66838  . In that sense, this PR sort of competes with https://github.com/llvm/llvm-project/pull/95328 , though it does have standalone merits.

This PR is meant for commits starting at [b503e21](https://github.com/llvm/llvm-project/pull/95840/commits/b503e2144a9845b2d5d9b1a554707a45acbb239d) . Depends on https://github.com/llvm/llvm-project/pull/91016

### CODE REVIEW

**Comment by jrbyrnes** [comment]
ping

File: llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp [hunk 1]
```diff
@@ -306,6 +306,18 @@ bool GCNTTIImpl::hasBranchDivergence(const Function *F) const {
   return !F || !ST->isSingleLaneExecution(*F);
 }
 
+unsigned GCNTTIImpl::getNumberOfParts(Type *Tp) const {
+  if (FixedVectorType *VTy = dyn_cast<FixedVectorType>(Tp)) {
+    if (DL.getTypeSizeInBits(VTy->getElementType()) == 8) {
+      unsigned ElCount = VTy->getElementCount().getFixedValue();
+      return ElCount / 4;
+    }
+  }
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:315]: This shouldn't need special casing?

File: llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp [hunk 2]
```diff
@@ -306,6 +306,18 @@ bool GCNTTIImpl::hasBranchDivergence(const Function *F) const {
...
+
+  std::pair<InstructionCost, MVT> LT = getTypeLegalizationCost(Tp);
+  return LT.first.isValid() ? *LT.first.getValue() : 0;
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:318]: This is really generic, why does this need overriding?

File: llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp [hunk 3]
```diff
@@ -337,9 +349,11 @@ unsigned GCNTTIImpl::getMinVectorRegisterBitWidth() const {
 unsigned GCNTTIImpl::getMaximumVF(unsigned ElemWidth, unsigned Opcode) const {
   if (Opcode == Instruction::Load || Opcode == Instruction::Store)
     return 32 * 4 / ElemWidth;
-  return (ElemWidth == 16 && ST->has16BitInsts()) ? 2
-       : (ElemWidth == 32 && ST->hasPackedFP32Ops()) ? 2
-       : 1;
+
+  return (ElemWidth == 8)                              ? 4
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:353]: Can you get away with just this part?

Comment (jrbyrnes) [reply llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:353]: Not really, no -- 

SLP will only attempt to vectorize candidates which have a vectorization factor greater than the number of parts (https://github.com/llvm/llvm-project/blob/ffca4ef5b1a8eff6097454df4b0f212e2393e41e/llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp#L16246). So, while we could theoretically increase the maximum VF to 8 or 16, this would only vectorize shuffles of size 8 or 16, and miss vectorization on the most important case of v4i8.

File: llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp (hunk 2 above)
Comment (jrbyrnes) [reply llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:318]: We are overriding the default (which just returns the legalization costs) because for certain instruction (e.g. v_perm) a v4i8 corresponds with 1 part. This enables SLP to consider i8 vectorization, and for most instructions the cost of vectorization is still the legalization cost.

Re generic: AFAICT the users are just vectorizers, which is what we want.

Comment (arsenm) [reply llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:318]: Then can you add a comment explaining this is hacking the legalization cost, and also implement the default case as calling the base implementation?

File: llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp (hunk 3 above)
Comment (arsenm) [reply llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:358]: Did you have a patch where you tried to make v4i8 legal somewhere?

Comment (jrbyrnes) [reply llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:358]: Only a local branch -- when I was working on it (a long while ago) I couldn't find a way around making i8s legal, but I could always take a look. This whole chain of work is a workaround for v4i8 not being legal, and there are still other areas that need work arounds -- mfmas using i32s in place of v4i8 and the weird isel patterns that causes come to mind.

Even if we were to reinvestigate that though, I think we should implement the workaround and delete later.

**Comment by jrbyrnes** [comment]
Ping, we would like to get this in by end of tomorrow to be on time for delivery

**Review by arsenm — COMMENTED** [review]
Can you rebase this and point me to the code in the vectorizer where this is making a difference? The SLP vectorizer's logic about what is vectorizable based on operation legality never really made sense to me.

**Comment by jrbyrnes** [comment]
> code in the vectorizer where this is making a difference? 


Sure -- for this work, there are two main pieces: 1. allowing SLP to consider i8s for vectorization, and 2. preferring vector shuffles for i8

1.

This is accomplished by changes to `getNumberOfParts` and `getMaximumVF`.

The most fundamental use of `TTI.getNumberOfParts` in SLP is in `tryToVectorizeList` ( https://github.com/llvm/llvm-project/blob/3c54aa14aa5f92ea2c96b85efbc7945ed55451e4/llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp#L16884 ). For early exit, it compares the vector size (VF, upwardly bounded by `TTI.getMaximumVF` ) to the `TTI.getNumberOfParts` for the vector type. If these two values are the same, then any vectorization will be undone by scalarization due to type-legalization.

2.

This is accomplished by changes to `getShuffleCost`

The pieces for 1. only consider the type and do not consider the operation which is obviously important (e.g. v2i16s may be legal for a given target, but the target doesn't have a vectorized UDiv operation for i16s). Operation considerations are encoded into the cost model, which SLP considers when calculating the cost of vectorizing (https://github.com/llvm/llvm-project/blob/3c54aa14aa5f92ea2c96b85efbc7945ed55451e4/llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp#L9401 called from https://github.com/llvm/llvm-project/blob/3c54aa14aa5f92ea2c96b85efbc7945ed55451e4/llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp#L10626 ) by comparing the total operation cost of the vectorized version of the code relative to the scalarized version. If vectorization does not offer enough cost improvement, then we don't vectorize. The opcode switch in `getEntryCost` is my usual starting point for reading SLP's use of the cost model.

To be honest, `getShuffleCost` is one of the most commonly used cost queries in SLP, so it's a bit hard to give you an exact pointer for this PR. That said, some rather important invocations are: checking the `ShuffleCost` when comparing cost of scalarized vs vectorized insertelements  ( https://github.com/llvm/llvm-project/blob/3c54aa14aa5f92ea2c96b85efbc7945ed55451e4/llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp#L9721 ), and adding cost when we need to shuffle the vector between operations in a vector chain ( https://github.com/llvm/llvm-project/blob/3c54aa14aa5f92ea2c96b85efbc7945ed55451e4/llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp#L9459 ) or for external users ( https://github.com/llvm/llvm-project/blob/3c54aa14aa5f92ea2c96b85efbc7945ed55451e4/llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp#L10928 ).

File: llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp [hunk 4]
```diff
@@ -1134,14 +1153,15 @@ InstructionCost GCNTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
 
   Kind = improveShuffleKindFromMask(Kind, Mask, VT, Index, SubTp);
 
-  // Larger vector widths may require additional instructions, but are
-  // typically cheaper than scalarized versions.
-  unsigned NumVectorElts = cast<FixedVectorType>(VT)->getNumElements();
+  unsigned ScalarSize = DL.getTypeSizeInBits(VT->getElementType());
```
[the remark this answers is not in the document]
Comment (jrbyrnes) [reply llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:1156]: Separate the operation / type cost changes into different PRs? That's actually what I've tried to do -- it's a PR stacking problem.

The relationship is mapped in the description, but basically this PR is meant for the changes to shufflecost and https://github.com/llvm/llvm-project/pull/91016 is meant for the changes required to enable SLP vectorization for i8s. The dependency is needed one way or the other to add the lit changes, and I decided to make https://github.com/llvm/llvm-project/pull/91016 the base as there is another PR in flight which depends on that as well.

In short, this PR is meant to address the shuffelcost changes and will be landed atomically with https://github.com/llvm/llvm-project/pull/91016  (if at all)

**Comment by jrbyrnes** [comment]
Force push for rebase

**Comment by jrbyrnes** [comment]
Rebase broke the history -- going to abandon and reopen..
